### PR TITLE
Move all persistence stuff to adapters

### PR DESCRIPTION
**Pull request by EPecherkin**
Adapter definition syntax:
```ruby
ActiveData.persistence_adapter('ClassName') do |data_source, primary_key, scope_proc|
  CustomAdapter.new(data_source, primary_key, scope_proc) # or anything that have similar interface to ActiveData::Model::Associations::PersistenceAdapters::Base
end
```
Where  
`ClassName` - name of model class or one of ancestors  
`data_source` - name of data source class  
`primary_key` - key to search data  
`scope_proc` - additional proc for filtering  

All required interface for adapters described in `ActiveData::Model::Associations::PersistenceAdapters::Base`.

Adapter for ActiveRecord is `ActiveData::Model::Associations::PersistenceAdapters::ActiveRecord`.
```ruby
ActiveData.persistence_adapter('ActiveRecord::Base') do |data_source, primary_key, scope_proc|
  PersistenceAdapters::ActiveRecord.new(data_source, primary_key, scope_proc)
end
```
So, all AR models will use `PersistenceAdapters::ActiveRecord` by default.

### CODE REVIEW

**Comment by alsemyonov** [comment]
Please, fix a typo (from “persistance” to “persistence”) in PR description and commit message.

**Comment by pyromaniac** [comment]
1. We probably don't need a default adapter, it doesn't make sense. We have to have a default adapter for 'ActiveRecord::Base'.
2. It would be better to use procs for adapter definitions like it is done for typecasters: https://github.com/pyromaniac/active_data/blob/master/lib/active_data.rb#L45
3. Adapter per association doesn't make sense, you will be unable to use it in tests, you are not going to redefine associations there. Instead, you are going to redefine adapters in the repository for the test environment.

**Comment by EPecherkin** [comment]
done

File: lib/active_data/model/associations/persistence_adapters/active_record.rb [hunk 1]
```diff
@@ -0,0 +1,51 @@
...
+
+      def delegate_to_original_scope?(method)
+        METHODS_EXCLUDED_FROM_DELEGATION.exclude?(method) && original_scope.respond_to?(method)
+      end
+    end
+
+    def scope(source)
+      scope = (scope_proc ? klass.unscoped.instance_exec(&scope_proc) : klass.unscoped)
+      scope = scope.where(primary_key => source)
+      ScopeProxy.new(scope)
```
Comment (pyromaniac) [inline lib/active_data/model/associations/persistence_adapters/active_record.rb]: Let's maybe initialize this in find_all? Also seems like find_all implementation screws the proxy up returning an array.

File: lib/active_data/model/associations/persistence_adapters/active_record.rb [hunk 2]
```diff
@@ -0,0 +1,51 @@
+module ActiveData::Model::Associations::PersistenceAdapters
```
Comment (pyromaniac) [inline lib/active_data/model/associations/persistence_adapters/active_record.rb]: module ActiveData
  module Model
    module Associations
      module PersistenceAdapters

File: Gemfile [hunk 3]
```diff
@@ -3,6 +3,6 @@ source 'https://rubygems.org'
 gemspec
 
 group :test do
-  gem 'guard'
-  gem 'guard-rspec'
+  gem 'guard', '~> 1.8.3'
```
Comment (pyromaniac) [inline Gemfile]: Remove this locks

File: lib/active_data/model/associations/embeds_one.rb [hunk 4]
```diff
@@ -48,7 +48,9 @@ def load_target
         def default
           return if evar_loaded?
 
-          default = reflection.default(owner) or return
+          default = reflection.default(owner)
```
Comment (pyromaniac) [inline lib/active_data/model/associations/embeds_one.rb]: wwwwhhhyyyyyy

File: lib/active_data.rb [hunk 5]
```diff
@@ -127,9 +129,15 @@ def self.config
...
   end
-end
 
-require 'active_data/model'
+  persistence_adapter('ActiveRecord::Base') do |data_source, primary_key, scope_proc|
+    ActiveData::Model::Associations::PersistenceAdapters::ActiveRecord.new(data_source, primary_key, scope_proc)
+  end
+
+  persistence_adapter('ActiveData::Model') do |data_source, primary_key, scope_proc|
+    ActiveData::Model::Associations::PersistenceAdapters::ActiveRecord.new(data_source, primary_key, scope_proc)
```
Comment (pyromaniac) [inline lib/active_data.rb]: Are you sure?

File: lib/active_data/model/associations/embeds_one.rb (hunk 4 above)
Comment (EPecherkin) [inline lib/active_data/model/associations/embeds_one.rb]: discussed. decided to keep like this for now

File: lib/active_data.rb (hunk 5 above)
Comment (EPecherkin) [inline lib/active_data.rb]: discussed. removed

**Pull request merged**